core-common/src/main/java/org/zalando/nakadi/service/KPIEventMapper.java

```diff
@@ -0,0 +1,109 @@
+package org.zalando.nakadi.service;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.generic.GenericRecordBuilder;
+import org.json.JSONObject;
+import org.zalando.nakadi.domain.kpi.KPIEvent;
+import org.zalando.nakadi.exceptions.runtime.InternalNakadiException;
+import org.zalando.nakadi.exceptions.runtime.NakadiBaseException;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class KPIEventMapper {
+
+    public static class KPIEventMappingException extends NakadiBaseException {
+        public KPIEventMappingException(final String message, final Throwable cause) {
+            super(message, cause);
+        }
+    }
+
+    private static class KPIGetter {
+        final String name;
+        final Method getter;
+
+        KPIGetter(final String name, final Method getter) {
+            this.name = name;
+            this.getter = getter;
+        }
+    }
+
+    private final Map<Class<? extends KPIEvent>, Set<KPIGetter>> fieldGetterCache = new HashMap<>();
+
+    public KPIEventMapper() {
+    }
+
+    public KPIEventMapper(final Set<Class<? extends KPIEvent>> mappedClasses) {
+        for (final Class<? extends KPIEvent> mappedClass : mappedClasses) {
+            fieldGetterCache.put(mappedClass, findGettersOfKPIFields(mappedClass));
+        }
+    }
+
+    private Set<KPIGetter> findGettersOfKPIFields(final Class<? extends KPIEvent> mappedClass) {
+        return Arrays.stream(mappedClass.getDeclaredFields())
+                .filter(field -> field.isAnnotationPresent(KPIEvent.KPIField.class))
+                .map(field -> findGetterByField(mappedClass, field))
+                .collect(Collectors.toSet());
+    }
+
+    private KPIGetter findGetterByField(final Class<? extends KPIEvent> clazz, final Field field) {
+        final var kpiField = field.getAnnotation(KPIEvent.KPIField.class);
+        final String getterName;
+        if (kpiField.getter().isEmpty()) {
+            final var fieldName = field.getName();
+            getterName = "get" + fieldName.substring(0, 1).toUpperCase() + fieldName.substring(1);
```

Comment:
this is quite dangerous. what about `isRunning()`, upper case... can you just get all the methods. 
for example, Jackson in some places makes [such checks](https://github.com/FasterXML/jackson-databind/blob/39885e702795dc4d57f416b3e44ea1f86d33d5af/src/main/java/com/fasterxml/jackson/databind/introspect/POJOPropertiesCollector.java#L680) to get differentiate getters / setters 

but ultimately it is better to use existing utils to access fields, getters etc. e.g. BeanUtils from Spring or Apache commons. however, I have limited experience writing such logic.

Reply:
I'll continue to improve this. I didn't want to over engineer with all the cases in the first version itself; especially without an example.